rest_tools/client/device_client.py

```diff
@@ -131,7 +119,85 @@ def DeviceGrantAuth(
             raise RuntimeError('Device authorization failed') from exc
         break
 
-    refresh_token = req['refresh_token']
+    return req['refresh_token']
+
+
+def DeviceGrantAuth(
+    address: str, token_url: str, client_id: str,
+    client_secret: Optional[str] = None,
+    scopes: Optional[List[str]] = None,
+) -> OpenIDRestClient:
+    """A REST client that can handle OpenID and the OAuth2 Device Client flow.
+
+    Args:
+        address (str): base address of REST API
+        token_url (str): base address of token service
+        client_id (str): client id
+        client_secret (str): client secret (optional - required for confidential clients)
+        scopes (list): token scope list (optional)
+    """
+    logger = logging.getLogger('DeviceGrantAuth')
+
+    auth = OpenIDAuth(token_url)
+    if 'device_authorization_endpoint' not in auth.provider_info:
+        raise RuntimeError('Device grant not supported by server')
+    endpoint = auth.provider_info['device_authorization_endpoint']
+
+    refresh_token = _perform_device_grant(logger, endpoint, auth.token_url, client_id, client_secret, scopes)
 
     return OpenIDRestClient(address=address, token_url=token_url, client_id=client_id,
                             client_secret=client_secret, refresh_token=refresh_token)
+
+
+def _load_token_from_file(filename: str) -> Optional[str]:
+    f = Path(filename)
+    if f.is_file():
+        try:
+            return f.read_text()
+        except Exception:
+            pass
+    return None
+
+
+def _save_token_to_file(filename: str, token: str) -> None:
+    f = Path(filename)
+    f.write_text(token)
+
+
+def SavedDeviceGrantAuth(
+    address: str, token_url: str,
+    filename: str,
```

Comment:
This could be a `Path` type. Relatedly, `filename=""` will err when you try to `write_text()`

Reply:
It could be. I'd really like to take a string to be easy on "dumb" users, and not force them to know about `pathlib`.  I guess I could take both?

Reply:
It's fine either way, I see your reasoning

Reply:
At the top, I'd use `filename = Path(filename)`. That saves a couple `Path()`-conversions.